packages/cli/src/cli/overview/action.create.overview.ts

```diff
@@ -0,0 +1,176 @@
+    });
+  }
+
+  protected onDefineParameters(): void {
+    this.source = this.defineStringParameter({
+      argumentName: 'SOURCE',
+      parameterLongName: '--source',
+      description: 'Path of source imagery files',
+      required: true,
+    });
+    this.maxZoom = this.defineIntegerParameter({
+      argumentName: 'MAX_ZOOM',
+      parameterLongName: '--max-zoom',
+      description: 'Output of the bundle file',
+    });
+    this.output = this.defineStringParameter({
+      argumentName: 'OUTPUT',
+      parameterLongName: '--output',
+      description: 'Path of output tar and index file',
+    });
+  }
+
+  async onExecute(): Promise<void> {
+    const logger = LogConfig.get();
+    const source = this.source.value;
+    if (source == null) throw new Error('Please provide a path for the source imagery.');
+    const maxZoom = this.maxZoom.value ?? DefaultMaxZoom;
+    const hash = createHash('sha256').update(source).digest('hex');
+    const path = fsa.join('overview', hash);
+
+    logger.info({ source, path }, 'CreateOverview: ListTiffs');
+    const tiffList = (await fsa.toArray(fsa.list(source))).filter(filterTiff);
+    const tiffSource = tiffList.map((path: string) => fsa.source(path));
+
+    logger.info({ source, path }, 'CreateOverview: prepareSourceFiles');
+    const tileMatrix = await this.prepareSourceFiles(tiffSource);
+
+    logger.info({ source, path }, 'CreateOverview: PrepareCovering');
+    const cutline = new Cutline(tileMatrix);
+    const builder = new CogBuilder(tileMatrix, MaxConcurrencyDefault, logger);
+    const metadata = await builder.build(tiffSource, cutline);
+
+    logger.info({ source, path }, 'CreateOverview: prepareTiles');
+    this.prepareTiles(metadata.files, maxZoom);
+    if (this.tiles.size < 1) throw new Error('Failed to prepare overviews.');
+
+    logger.info({ source, path }, 'CreateOverview: GenerateTiles');
+    await this.generateTiles(path, tileMatrix);
+
+    logger.info({ source, path }, 'CreateOverview: CreatingTarFile');
+    await this.createTar(path, logger);
+
+    logger.info({ source, path }, 'CreateOverview: Finished');
+    return;
+  }
+
+  async prepareTiles(files: NamedBounds[], maxZoom: number): Promise<void> {
+    for (const file of files) {
+      const name = file.name;
+      const [z, x, y] = path.basename(name).replace('.tiff', '').split('-').map(Number);
+      let qk = QuadKey.fromTile({ x, y, z });
+      this.addChildren(qk, maxZoom);
+      while (qk.length > 0) {
+        if (this.tiles.has(qk)) break;
+        this.tiles.add(qk);
+        qk = QuadKey.parent(qk);
+      }
+    }
+  }
+
+  addChildren(qk: string, maxZoom: number): void {
+    if (qk.length >= maxZoom) return;
+    for (const child of QuadKey.children(qk)) {
+      this.tiles.add(child);
+      if (child.length < maxZoom) this.addChildren(child, maxZoom);
+    }
+  }
+
+  async prepareSourceFiles(sources: ChunkSource[]): Promise<TileMatrixSet> {
```

Comment:
why are we reading in all files to just get the tile matrix?

Reply:
This is mainly used for preparing `this.files: NamedBounds[] = [];`, and just getting the tileMatrix at the same time.
`this.file` will need to pass in to tile-generator for reference the source files and find bounds intersection with tiles.

Reply:
the metadata builder also reads in all the files? doesnt it also create a file list?

I think having these functions modify `this.` state makes it somewhat hard to follow, these could be functions that just return the data?

Reply:
It is the different files list from metatdata builder, it is the cogs not the source files. I am prepare a new files with the bounds and sources. 

Cool, I can update the function to return a tilematrix and files instead of using this.files.